### ISSUE TITLE
[Umbrella] Implement QueueingHintFn in in-tree plugins

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#118893]
### What would you like to be added?

In https://github.com/kubernetes/kubernetes/pull/118551, we implemented `QueueingHintFn` in `EventsToRegister ` which allows each plugin to filter out useless events to requeue Pod back to the activeQ. (See the PR or [the original proposal](https://docs.google.com/document/d/1S1_OqwCN2oY0Xp4dZt6XLCQQGu7dWm96blgXmGKhwSo/edit#heading=h.xpaj7b9yvnh3) about the details)
Basically, all in-tree plugins, which implements PreFilter or Filter, should have `QueueingHintFn` to achieve the efficient requeueing. 

Please feel free to assign what you'd like to work on. And please comment in this issue before starting to work on so that we can avoid many people from working on the same thing!

- [x] InterPodAffinity @nayihz https://github.com/kubernetes/kubernetes/pull/122471
- [x] Dynamic resource @pohly https://github.com/kubernetes/kubernetes/pull/119078
- [x] NodeAffinity @carlory https://github.com/kubernetes/kubernetes/pull/119155 https://github.com/kubernetes/kubernetes/pull/122309
- [x] NodePorts @carlory https://github.com/kubernetes/kubernetes/pull/119176
- [x] NodeResourceFit  @carlory https://github.com/kubernetes/kubernetes/pull/119177
- [x] NodeUnschedulable @wackxu @carlory  https://github.com/kubernetes/kubernetes/pull/119396 https://github.com/kubernetes/kubernetes/pull/122334
- [x] CSILimits @utam0k https://github.com/kubernetes/kubernetes/pull/121508
- [x] TopologySpread @nayihz https://github.com/kubernetes/kubernetes/pull/122195
- [x] TaintToleration  @wackxu @sanposhiho https://github.com/kubernetes/kubernetes/pull/119397 https://github.com/kubernetes/kubernetes/pull/124287
- [ ] VolumeBinding @y-taka-23 
- [x] VolumeRestriction @HirazawaUi https://github.com/kubernetes/kubernetes/pull/125279 https://github.com/kubernetes/kubernetes/pull/125280
- [x] VolumeZone @Gekko0114 https://github.com/kubernetes/kubernetes/pull/119373

/sig scheduling
/triage accepted
/priority important-longterm

### Why is this needed?

It will contribute to the enqueue efficiency a lot and thus overall scheduling performance should get improved significantly by preventing the useless retry on the scheduling as much as possible.

**Comment by sanposhiho** [comment]
@pohly I assigned DRA to you 🙇 Could you raise a PR for DRA like https://github.com/kubernetes/kubernetes/pull/117844 implements with adding some unit tests? That'll be blocked by https://github.com/kubernetes/kubernetes/pull/118529 eventually though.

**Comment by carlory** [comment]
Hi @sanposhiho , I would like to work on this issue. I'll pick up the NodeAffinity, NodePorts and NodeResourceFit.

**Comment by wackxu** [comment]
I'd like to work on NodeUnschedulable，TaintToleration
/assign

**Comment by czybjtu** [comment]
I'll pick up the TopologySpread
/assign

**Comment by sanposhiho** [comment]
@czybjtu @wackxu @carlory 
Thanks, I assigned them to you all. Please ping me in your PRs so that I can give some looks. 
Also, feel free to reach out me if you have any questions around (either here or the k8s slack), I know it's kind of difficult to fully understand because there is actually no place yet using QueueingHintFn.

**Comment by utam0k** [comment]
I am interested in it. assign me.

**Comment by sanposhiho** [comment]
/assign @utam0k 

Which one would you want to pick up?

**Comment by utam0k** [comment]
> /assign @utam0k 
> 
> Which one would you like to pick up?

Anything, but I'm happy if it's a plugin I haven't touched other issues.

**Comment by sanposhiho** [comment]
I assigned CSILimits/nonCSILimits to you.

**Comment by y-taka-23** [comment]
Hello @sanposhiho, could you assign me too if you don't mind?

**Comment by utam0k** [comment]
@sanposhiho :pray: :+1:

**Comment by sanposhiho** [comment]
@y-taka-23 assigned VolumeBinding to you!

**Comment by y-taka-23** [comment]
@sanposhiho Thx!

**Comment by Gekko0114** [comment]
Hi @sanposhiho ,

If there are any remaining items, could you assign it to me?

**Comment by HirazawaUi** [comment]
I'd like to work for VolumeRestriction
/assign

**Comment by sanposhiho** [comment]
@Gekko0114 @HirazawaUi I assigned VolumeRestriction and VolumeZone to you both

**Comment by Gekko0114** [comment]
Sure, thank you so much!

**Comment by z1cheng** [comment]
Hey @sanposhiho, could you please assign it to me if there are any remaining items? :)

**Comment by sanposhiho** [comment]
@z1cheng Unfortunately, no more slots here.

**Comment by pohly** [comment]
> I assigned DRA to you 🙇 Could you raise a PR for DRA like https://github.com/kubernetes/kubernetes/pull/117844 implements with adding some unit tests?

See https://github.com/kubernetes/kubernetes/pull/119078

> That'll be blocked by https://github.com/kubernetes/kubernetes/pull/118529 eventually though.

Are there any negative consequences when adding hints without that PR? Tests seem to be passing without 118529.

> I know it may be kind of difficult to fully understand the concept because there is actually no place yet using QueueingHintFn.

I suppose DRA in 119078 can serve as example. For example, note how it checks for `cache.DeletedFinalStateUnknown` - that is often something that gets forgotten when dealing with informer callbacks.

**Comment by sanposhiho** [comment]
> Are there any negative consequences when adding hints without that PR? Tests seem to be passing without 118529.

When the Pod is rejected by the DRA `Reserve()`, the scheduling queue doesn't call DRA's hint fn without 118529. 
Yep, on second thought, that affects badly only the performance, not existing UTs or integrate tests.

**Comment by sanposhiho** [comment]
We're working on it (seems yours slipped though).

**Comment by sanposhiho** [comment]
One semantic change to QueueingHint will soon be merged. Please check it out, everyone. 
https://github.com/kubernetes/kubernetes/pull/119290

**Comment by sanposhiho** [comment]
FYI: We ended up having KEP for QueueingHint.
https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/4247-queueinghint/README.md
The upcoming semantic change https://github.com/kubernetes/kubernetes/pull/119517/ is explained there.

**Comment by utam0k** [comment]
I'm working on `CSILimits`. I'd like to confirm QHint direction for it. For example, if the PVCs associated with the deleted Pod match the PVCs of the Pod waiting in the queue (i.e. the driver matches), then we can return some hints, right? I'm sure there are other times when CSINode is added, but for now I just want to confirm that this is what it looks like.

**Comment by alculquicondor** [comment]
For anyone writing these PRs: please put a release note.
While the change doesn't have an API, it has a behavior change that could be user-visible.
See  https://github.com/kubernetes/kubernetes/pull/119155 for an example.

**Comment by sanposhiho** [comment]
- https://github.com/kubernetes/kubernetes/pull/122285
- https://github.com/kubernetes/kubernetes/pull/122288

The ones for NodeUnschedulable and NodeAffinity are once reverted to have a cherry-pick.

https://github.com/kubernetes/kubernetes/pull/122309 is the new one for NodeAffinity.
@wackxu Please submit a new one for NodeUnschedulable if you have time. (If not, we can ask someone to take over

**Comment by carlory** [comment]
@sanposhiho I sumbit a pr to reimplement it and add @wackxu as co-author. If @wackxu has been still working on this task, feel free to close [this pr](https://github.com/kubernetes/kubernetes/pull/122334).

**Comment by sanposhiho** [comment]
After implementing QueueingHint, I'd like to ask you to implement an integration test for the queueing hint if you're interested in: https://github.com/kubernetes/kubernetes/issues/122305

**Comment by carlory** [comment]
> Everyone, after implementing QueueingHint, I'd like to ask you to implement an integration test for the queueing hint if you're interested in: #122305

I can try it, but I doesn't know how to write a suitable test for this feature. Perhaps I need your help during implementation.

BTW @sanposhiho I want to confirm whether the nodename plugin should implement the feature? if needed, I can do it.

**Comment by sanposhiho** [comment]
> I can try it, but I don't know how to write a suitable test for this feature. Perhaps I need your help during implementation.

Thanks. Please feel free to open a PoC (= small) PR and ping me there. We can discuss, decide on the direction, and build it together there.

>  I want to confirm whether the nodename plugin should implement the feature? if needed, I can do it.

No need for nodename. It's actually unused.

**Comment by AxeZhan** [comment]
@y-taka-23 Are you still working on volumebinding? I can take it over if you're overloaded :)

**Comment by Gekko0114** [comment]
> Everyone, after implementing QueueingHint, I'd like to ask you to implement an integration test for the queueing hint if you're interested in: #122305

Hi @sanposhiho ,
Sure, and I have two questions.
Q1: Are there any example codes for integration tests?
Q2: I already created a PR for queueing hint. Should I add integration test to an existing PR or should I create a new PR after completing current my PR?

**Comment by sanposhiho** [comment]
> Q1: Are there any example codes for integration tests?

Not yet. We have to craft from scratch. My basic idea for that is https://github.com/kubernetes/kubernetes/issues/122305#issuecomment-1858811508. 

> Q2: I already created a PR for queueing hint. Should I add integration test to an existing PR or should I create a new PR after completing current my PR?

Well, either is fine though, I _prefer_ doing it in another PR because UT should be covering basic scenarios and thus the integration tests can be follow-up. (Plus, your PR is especially big already.)

**Comment by sanposhiho** [comment]
Also is anyone willing to take over InterPodAffinity, which is currently assigned to me? (I'm pretty overloaded now. 😓 )

**Comment by nayihz** [comment]
> Also is anyone willing to take over InterPodAffinity, which is currently assigned to me? (I'm pretty overloaded now. 😓 )

I'd like to work on this. Pls assign to me.

**Comment by AxeZhan** [comment]
> @y-taka-23 Are you still working on volumebinding? I can take it over if you're overloaded :)

Kindly ping @y-taka-23

**Comment by y-taka-23** [comment]
@AxeZhan (Cc: @Gekko0114) Sorry for the late reply. Can I continue to work on VolumeBinding for some more time?

**Comment by sanposhiho** [comment]
@y-taka-23 
From our experience so far, I'd recommend splitting PRs for each event like @utam0k is doing to make each of them smaller. Especially, VolumeBinding is subscribing many events relatively.

**Comment by alculquicondor** [comment]
@Huang-Wei and I were discussing the topic of SchedulingHints a little.

The hints seem to be very verbose into all the possibilities that can make a pod schedulable, which can lead to:
- a high maintenance cost
- expensive event handlers, ultimately affecting overall performance

OTOH, we have seen in our fleet that the SchedulingHints framework, without any hints enabled, lead to high memory utilization (50% increase for the fleet) and, in big clusters, patterns of memory leaks.

With all of that, it's becoming hard to justify further investments in the feature, at least in its current form.

Listing all positive alternatives:
- Remove the feature (what are we losing?)
- Rework the framework, how events are stored, maybe put caps, and so on.
- Limit the complexity that scheduling hints are allowed to have (for example, disallow querying of multiple objects to make a decision).

(I wanted to discuss this in a SIG meeting, but it's hard due to the timezone difference with @sanposhiho).

FYI @pohly, as I know DRA kind of depends on this feature. But DRA could implement more traditional (non-extensible) event handlers. Question: do scheduling hints matter equally for the semantic model?

**Comment by sanposhiho** [comment]
Let me understand the problem first.

> a high maintenance cost

This is.. Yea, many PRs are waiting for reviews while our bandwidth is very limited. And also we'll have to keep them maintained in the future too.

> expensive event handlers, ultimately affecting overall performance

Any GitHub issue we have? I couldn't find and don't remember a report of such.

> OTOH, we have seen in our fleet that the SchedulingHints framework, without any hints enabled, lead to high memory utilization (50% increase for the fleet) and, in big clusters, patterns of memory leaks.

We noticed it, and have possible solutions -
https://github.com/kubernetes/kubernetes/issues/120622

and... reviews are pending over there too, though.

> Listing all positive alternatives:

So, if the core problem is that we cannot spare any more time on this feature (esp reviews), then the only complete solution would be the removal of the feature, unfortunately. I don't quite understand the last one; any QHints that are too complex actually? 

Regarding the memory issue (and possibly throughput issue(?) that I haven't grasped yet), there are solutions we haven't yet done, and "Rework the framework, how events are stored" should be considered after them (if they don't solve the issue enough).

**Comment by pohly** [comment]
> FYI @pohly, as I know DRA kind of depends on this feature. But DRA could implement more traditional (non-extensible) event handlers. Question: do scheduling hints matter equally for the semantic model?

They probably matter less because there's less back-and-forth through the apiserver. But there are still some cases where specific pods get put into the unschedulable queue and become schedulable only when specific claims are updated. To give you a full answer I'll have to set up a benchmark in scheduler_perf and do a comparison with and with queuing hints - I'll add this to my TODO list...

**Comment by kerthcet** [comment]
What I learned about queueingHint from code reading and reviewing is a) the hint mechanism itself is not that complex (except the inflight pods), thanks to @sanposhiho efforts, but the hintFunc is trivial 2) we do not have a report or test about how hint improves the scheduling throughout although we always said that theoretically it does. 3) queueingHint does solve some problems like `preCheck` and inflight pods 4) We should rework about the memory leakage problem, but actually this is a problem we have now for missing the inflight pods.

TL;DR: I can take over the review work once we reach an agreement. Like whether it's possible to fix the memory problem in prior, how to make the hintFunc simple, IIUC, if we do nothing, we still have the similar performance as we have before.

**Comment by alculquicondor** [comment]
cc @gabesaba

**Comment by alculquicondor** [comment]
>> expensive event handlers, ultimately affecting overall performance
>
> Any GitHub issue we have? Which performance is it affecting? I couldn't find and don't remember a report of such.

I didn't see any issue specific to scheduling hints, however @gabesaba just filed #124384, in which calculations from the traditional event handlers hold the lock for too long, dramatically increasing the scheduling latency. Making complex scheduling hints has the potential of running into similar problems.

> We noticed it, and have possible solutions -
> https://github.com/kubernetes/kubernetes/issues/120622
>
> and... reviews are pending over there too, though.

Yes, I acknowledge that, but it's not just about reviews. We don't currently have reproducible scenarios that can tell us if the solution will scale. A reviewer might not be able to see this. I'm hoping we can enhance the scheduler_perf tests, bringing them to the extreme. Without them, it's hard to discern whether a change is good.

> I don't quite understand the last one; any QHints that are too complex actually? If yes, we can loosen the scope of those specific QHints only. I'm not sure we have to have some such limitations entirely.

What I mean is that the code of each of the scheduling hints is quite complex. This means that we could easily leave pods unschedulable if we missed a case or we introduce new filtering behavior and we forget to update the hints accordingly. But also runtime complexity. My proposal would be along the lines of keeping all scheduling hints around 10 lines max, and avoiding loops over lists of objects. Basically, we should be doing O(1) operations on the object update.

Back to reviews.... it comes back to the same complexity point above. If the hint is too complex, it's hard for a reviewer to keep a mental model of all the possible scenarios that could make a pod schedulable, and to be sure that nothing was missed. Maybe we can start with simpler checks that could have false positives, because today almost every event would make a pod go back to the scheduling queue, so it would not be a regression.

**Comment by sanposhiho** [comment]
After reading all the points from @alculquicondor, 

I said:
> if the core problem is that we cannot spare any more time on this feature (esp reviews) based on priorities, then the only complete solution would be the removal of the feature, unfortunately. 
https://github.com/kubernetes/kubernetes/issues/118893#issuecomment-2062867701

But, seems like it's not the case. And, @kerthcet raised a hand for help of reviewing (thanks a lot!), which would improve the situation where we only had me continuously taking the review effort for QHints, and PRs got eventually blocked for the second review.

Regarding two issues (high memory consumption and throughput), definitely we have to work on them, including scheduler_perf improvement to measure it. (all action items are already listed in https://github.com/kubernetes/kubernetes/issues/120622) 
However, again, we have done _nothing_ literally for those issues; it's too early to say we cannot solve them technically. 

Lastly about the complexity of QHint,

> My proposal would be along the lines of keeping all scheduling hints around 10 lines max, and avoiding loops over lists of objects. Basically, we should be doing O(1) operations on the object update.

I'd agree with having "avoiding loops" policy, which I also believe we can avoid one kind of complexity.

But, disagree with 10 lines max; Regardless of the line length, it should be accepted as long as it's simple enough, which I believe most of the current QHints are. 

I admit https://github.com/kubernetes/kubernetes/pull/121508 is very complex and probably we can reduce the scope. 
But, I don't remember other QHints which I feel too complex to be manageable. ...Yes, it's subjective, maybe some other QHints look very complex to some people, so, I asked "which one(s) is too complex" at the first comment. We can reduce the scope of them accordingly so that we can make them simple enough / maintainable.

**Comment by alculquicondor** [comment]
> But, disagree with 10 lines max;

I just meant it as a general rule of thumb. But I think you got the idea.

Not sure if I would call this one too complex, but it accesses a lister.
https://github.com/kubernetes/kubernetes/blob/5e229a02d34976bf41ad93d2a8ad86ac28ac5f84/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L437

In the topic of reviews, I'm trying to make some time to review more PRs, but it's been challenging. I hope to have more time next month.

**Comment by sanposhiho** [comment]
Yes, let's confirm here that we can reduce the scope if QHint is growing too complex. 
One criterion is "avoiding loops over lists of objects", but the reviewers can ask to reduce the scope in other cases too; if one out of two reviewers feels the complexity, then that's enough sign to reduce the scope.

We agree to reduce the scope of https://github.com/kubernetes/kubernetes/pull/121508. (cc @utam0k) 
I'll revisit all the other QHints (including pending PRs) later for this perspective. 

I'll defer the decision of DRA one to @pohly ([ref](https://github.com/kubernetes/kubernetes/issues/118893#issuecomment-2066524982)).

**Comment by utam0k** [comment]
I also noticed that admin #12150 is too complex to maintain. Let me reduce the scope of #121508 . How can we decide the criterion for QHint scope?

Also, I have already created and submitted the PRs for QHint. I'd be happy if I could help with the reviews for PRs related to QHint.  If you assign me as the first reviewer, I may be able to comment on the sections that obviously need comment. This would surely reduce the review burden.

**Comment by alculquicondor** [comment]
> If you assign me as the first reviewer, I may be able to comment on the sections that obviously need comment. This would surely reduce the review burden.

Thanks for stepping up. Always feel free to take the first passes. Unfortunately, we are rather limited on the time to get a second pass from approvers. And that's another reason to start with simple, smaller PRs.

**Comment by alculquicondor** [comment]
That said, @sanposhiho has been doing tremendous work with reviews. Again, I'm trying to catch up, but I can't promise much yet.

**Comment by pohly** [comment]
Regarding DRA: I had already added one performance test for structured parameters earlier, about the common case with one claim per pod. `BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_200nodes` improves a bit when enabling queuing hints (BTW, is it queuing or queueing?).

Average throughput increases by +14.89%, average scheduling duration goes down by -17.06%.

The less interesting from a real-world relevance perspective case is multiple pods sharing the same claim. I expect a higher impact of queuing hints there, but I need to work a bit on scheduler_perf before I can set up such a workload there.

**Comment by alculquicondor** [comment]
`2000pods_200nodes`

I think problems only start to arise at 5k nodes with high churn (pods terminating and new pods incoming)

**Comment by pohly** [comment]
For "classic DRA" there was a significant slowdown when disabling queuing hints for `2000pods_200nodes`. I can re-run the structured parameters scenarios at higher pod and node counts.

I also have the variant with shared claims now (PR coming soon). Against my expectation, average throughput is about the same there. Where queuing hints help is for the `scheduler_scheduling_attempt_duration_ms` percentiles - `scheduler_scheduling_attempt_duration_ms/Perc99` gets reduced by 80.15%.

**Comment by pohly** [comment]
Does scheduler_perf simulate churn, i.e pods terminating? At least the DRA scenarios are all about "fill up the cluster" (= only concerned about starting as many pods as possible on the cluster).

**Comment by alculquicondor** [comment]
> Does scheduler_perf simulate churn, i.e pods terminating?

Maybe some tests do. But churn is actually were we found problems here #124384.

> Where queuing hints help is for the scheduler_scheduling_attempt_duration_ms percentiles - scheduler_scheduling_attempt_duration_ms/Perc99 gets reduced by 80.15%.

That sounds very surprising.... scheduling hints helps in the latency of a single scheduling_attempt?

**Comment by alculquicondor** [comment]
Thanks.

In any case, I think we are going down the path of keeping each plugin hint as simple as possible and trying to optimize the overall framework.

Maybe the hint for DRA will have a chance to simplify as we progress with the semantic model and focus on hints just to optimize that path.

**Comment by sanposhiho** [comment]
Regarding the throughput topic mentioned here, we discussed at https://github.com/kubernetes/kubernetes/issues/124384 and just created the one for tracking: https://github.com/kubernetes/kubernetes/issues/124566

**Comment by bells17** [comment]
@y-taka-23 (cc: @AxeZhan @Gekko0114 @sanposhiho) Regarding the implementation of the QueueingHintFn for VolumeBinding, if you are unable to secure time to work on it, would it be okay for me to implement it?

**Comment by sanposhiho** [comment]
@bells17 Feel free to proceed if no response from them for a while, e.g., after GW.
Also, if @AxeZhan is still interested in taking part of it, you both can discuss and decide which event to work on. (There're many events registered in VolumeBinding.)

**Comment by AxeZhan** [comment]
@bells17 Feel free to start working on it, I can help reviewing :)
Notice that the logic of `volume binding` is pretty complex(there involves finding pv,pvc and sc), and we may want to reduce the complexity in queueingHint functions.

**Comment by YamasouA** [comment]
Hi.
I am interested in implementing this feature.
I would like to participate this.

**Comment by sanposhiho** [comment]
@bells17 @YamasouA
Please leave a comment about which event you want to pick up before starting to prevent conflicting work.
Here's the list of events in VolumeBinding.
- [ ] StorageClass (Add/Update)
- [ ] PVC (Add/Update)
- [ ] PV (Add/Update)
- [ ] Node (Add/UpdateNodeLabel/UpdateNodeTaint)
- [ ] CSINode (Add/Update)
- [ ] CSIDriver (Add/Update)
- [ ] CSIStorageCapacity (Add/Update)
 
https://github.com/kubernetes/kubernetes/blob/1dc30bf90fd6a729d226b4e942118110b0a73e65/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L96-L121

**Comment by bells17** [comment]
StorageClass (Add/Update)
PVC (Add/Update)
PV (Add/Update)
CSIStorageCapacity (Add/Update)

**Comment by YamasouA** [comment]
I would like to pick up the other three events.

**Comment by Gekko0114** [comment]
@sanposhiho and others,
Sorry, I don't follow the discussion in detail.
I've created a PR #119373 several months ago, and I've rebased the PR because of conflict.
And I am waiting for an approval from sig-storage team members.

Should I change the scope of this PR? Or can I just wait for an approval?

**Comment by AxeZhan** [comment]
@Gekko0114 
Considering the context of #119373 , I think it's ok to keep it as it is now.
It's been a real long time since sig-storage reviewed ... Maybe you can try reach the slack channel of sig-storage.

**Comment by bells17** [comment]
I think there are changes that require sig-storage's review, including VolumeZone, VolumeRestriction, and possibly VolumeBinding, which is currently being implemented.
So, if you're planning to request a review via Slack, it might be better for communication to ask sig-storage to review these three PRs together.

@sanposhiho @HirazawaUi By the way, I noticed that the following PR hasn't been updated for a while. Could you give me an update on its current status?
If you're having trouble finding time to work on the changes in this PR due to other commitments, I'd be happy to take over. Would you be open to that?
https://github.com/kubernetes/kubernetes/pull/119405

**Comment by Gekko0114** [comment]
> I think there are changes that require sig-storage's review, including VolumeZone, VolumeRestriction, and possibly VolumeBinding, which is currently being implemented.
So, if you're planning to request a review via Slack, it might be better for communication to ask sig-storage to review these three PRs together.

Thank you for your suggestion! I already asked via slack here with @HirazawaUi several months ago. It is being stacked by sig-storage team. I will ping them again.
https://kubernetes.slack.com/archives/C09QZFCE5/p1706531933985289?thread_ts=1701594690.661529&cid=C09QZFCE5

**Comment by sanposhiho** [comment]
Let's narrow down the scope of each PR by, for example, creating a PR per event instead of per plugin so that people can take a look more easily. Also, additionally you can attend [their SIG meeting](https://github.com/kubernetes/community/blob/master/sig-storage/README.md) to ask for help directly (I know many people attending the discussion here would have a timezone problem though).

**Comment by carlory** [comment]
@sanposhiho I can help to review storage pr if I have time. please assign/cc it to me if it's ready for review.

**Comment by bells17** [comment]
volumebinding plugins PR(WIP):

- https://github.com/kubernetes/kubernetes/pull/124958
- https://github.com/kubernetes/kubernetes/pull/124959
- https://github.com/kubernetes/kubernetes/pull/124960
- https://github.com/kubernetes/kubernetes/pull/124961

https://github.com/kubernetes/kubernetes/issues/118893#issuecomment-2097473049

**Comment by Gekko0114** [comment]
I've divided volume_zone's PR into 4 parts.

https://github.com/kubernetes/kubernetes/pull/125000
https://github.com/kubernetes/kubernetes/pull/125001
https://github.com/kubernetes/kubernetes/pull/124996
https://github.com/kubernetes/kubernetes/pull/124998

**Comment by Gekko0114** [comment]
Completed volumezone scheduling, thanks @sanposhiho for all your supports!

**Comment by googs1025** [comment]
Is there any unfinished work on this issue? If so, I will take it.

**Comment by sanposhiho** [comment]
No, I believe all necessary QHints are implemented or already have opened PRs.

**Comment by sanposhiho** [comment]
Hey everyone. Thanks to everyone's effort, I believe we only have one remaining PR (https://github.com/kubernetes/kubernetes/pull/125171), and all the others are merged already.

Actually, we have another issue that aims to increase the integration test coverage for QueueingHints: https://github.com/kubernetes/kubernetes/issues/122305
Help us with this one too if you're interested in ;)
